### PR TITLE
DRILL-6540: Upgrade to HADOOP-3.x libraries

### PR DESCRIPTION
**Pull request by agozhiy**
Set in order changes from this PR: https://github.com/apache/drill/pull/1732

Updated Hadoop and HBase libraries to the latest versions:
Hadoop: 3.2.1
HBase: 2.2.2

Jira - [DRILL-6540](https://issues.apache.org/jira/browse/DRILL-6540).

### CODE REVIEW

**Comment by arina-ielchiieva** [comment]
By @vvysotskyi:

Here is the list of things to be done before merging this PR:

- [ ] Cleanup poms:
- - [ ] Remove commented out lines, revise new dependencies to reduce JDBC driver size
- - [ ] Revert removing excluding `commons-logging`, it will be better to add an absent mock class from `commons-logging` to Drill
- [ ] Remove change with logging test error in `ITTestShadedJar.java`
- [ ] Test Drill on Hadoop 3 cluster (secure, non-secure modes)
- [ ] Check with Drill-on-YARN
- [ ] Test Drill in embedded mode on Windows (test Hadoop win-utils)
- [ ] Test JDBC client with new Hadoop version
- [ ] Check whether new Hadoop version API is compatible with the current one, in the case if not, decide whether to introduce new profile, use a property to specify the version during building or document supported versions

File: exec/jdbc-all/src/test/java/org/apache/drill/jdbc/DrillbitClassLoader.java [hunk 1]
```diff
@@ -61,21 +58,21 @@ public DrillbitClassLoader() {
    *
    * Taken from Apache Harmony
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply exec/jdbc-all/src/test/java/org/apache/drill/jdbc/DrillbitClassLoader.java]: Agree, this method can be replaced

File: pom.xml [hunk 2]
```diff
@@ -2046,6 +2124,76 @@
...
+                <groupId>io.netty</groupId>
+                <artifactId>netty-all</artifactId>
+              </exclusion>
+              <exclusion>
+                <groupId>io.netty</groupId>
+                <artifactId>netty</artifactId>
+              </exclusion>
+            </exclusions>
+          </dependency>
+          <!-- Hadoop Test Dependencies -->
```
Comment (vdiravka) [inline pom.xml]:
```suggestion
          <!-- Hadoop Test Dependencies -->
          
```

File: exec/java-exec/src/test/java/org/apache/drill/exec/work/batch/FileTest.java [hunk 3]
```diff
@@ -43,7 +43,7 @@ public static void main(String[] args) throws IOException {
     FSDataOutputStream out = fs.create(path);
     byte[] s = "hello world".getBytes();
     out.write(s);
-    out.sync();
```
Comment (vdiravka) [inline exec/java-exec/src/test/java/org/apache/drill/exec/work/batch/FileTest.java]: minor:
`sync` -> `hflush`
https://github.com/mapr/hadoop-common/blob/release-2.7.0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Syncable.java#L31

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/LocalSyncableFileSystem.java [hunk 4]
```diff
@@ -141,7 +141,7 @@ public LocalSyncableOutputStream(Path path) throws FileNotFoundException {
       output = new BufferedOutputStream(fos, 64*1024);
     }
 
-    @Override
+    // TODO: remove it after upgrade MapR profile onto hadoop.version 3.1
     public void sync() throws IOException {
       output.flush();
       fos.getFD().sync();
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/LocalSyncableFileSystem.java]: `hsync` and `hflush` can be unified too

File: exec/java-exec/src/main/java/org/apache/commons/logging/impl/Log4JLogger.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.logging.impl;
+
+public class Log4JLogger {
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/commons/logging/impl/Log4JLogger.java]: https://issues.apache.org/jira/browse/HADOOP-15749

[the remark this answers is not in the document]
Comment (agozhiy) [reply exec/java-exec/src/main/java/org/apache/commons/logging/impl/Log4JLogger.java]: Done.

File: exec/java-exec/src/test/java/org/apache/drill/exec/work/batch/FileTest.java (hunk 3 above)
Comment (agozhiy) [reply exec/java-exec/src/test/java/org/apache/drill/exec/work/batch/FileTest.java]: Done.

File: pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (agozhiy) [reply pom.xml]: Done.

Comment (agozhiy) [reply pom.xml]: Done.

**Comment by agozhiy** [comment]
As it appeared, the new Hadoop library version required updating of hadoop-winutils, though there is  only an artifact for the version 2.7.0 in maven repository. So I created the needed artifact using jitpack.io.

**Comment by vvysotskyi** [comment]
@agozhiy, since winutils binaries are hosted in your repository, could you please add `README.md` file with the link to the repository and steps on how to publish new versions. As an example, may be used `Calcite.md`.

File: docs/dev/HadoopWinutils.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+## Hadoop Winutils
+
+Hadoop Winutils native libraries are required to run Drill on Windows. The last version presented in maven repository is 2.7.1 and doesn't updated anymore.
```
Comment (arina-ielchiieva) [inline docs/dev/HadoopWinutils.md]: presented in maven repository -> present in Maven repository
doesn't updated anymore -> is not updated

File: docs/dev/HadoopWinutils.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+So instead of heaving a dependency Drill now contains required files as resources at distribution/src/main/resources.
```
Comment (arina-ielchiieva) [inline docs/dev/HadoopWinutils.md]: That's why Winutils version matching Hadoop version used in Drill is located in distribution/src/main/resources.

File: docs/dev/HadoopWinutils.md (hunk 6 above)
Comment (agozhiy) [reply docs/dev/HadoopWinutils.md]: Fixed.

File: docs/dev/HadoopWinutils.md (hunk 7 above)
Comment (agozhiy) [reply docs/dev/HadoopWinutils.md]: Fixed, thanks.

**Comment by vvysotskyi** [comment]
There are a lot of unit tests failures with these changes, most of the errors are the following:
```
java.lang.NoSuchMethodError: com.google.common.base.Preconditions.checkState(ZLjava/lang/String;I)V
```
It looks like some tests which don't patch guava loads `Preconditions` class and then it cannot be patched in other classes where it is required. All unit tests classes should be revised and patching should be added where it was absent, so I propose to fix it in the scope of another Jira, but before merging this PR.

**Pull request closed without merging**